projects/laji/src/app/shared-modules/tree-select/tree-selector/tree-selector.component.ts

```diff
@@ -0,0 +1,297 @@
+})
+export class TreeSelectorComponent implements OnInit {
+  @Input() selectedOptions: SelectedOption[] = [];
+  @Input() optionsTree: TreeOptionsNode[];
+  @Input() multiselect = false;
+  @Input() tristate = false;
+  @Input() includeCheckbox = false;
+  @Input() includeCount = false;
+  @Input() includeLink = false;
+  @Input() includeQualityIcon = false;
+  @Input() openOnSelect = false;
+  @ViewChild('tree') treeComponent: TreeComponent;
+  @Output() emitSelect = new EventEmitter<SelectedOption[]>();
+
+  treeModel: TreeModel;
+  checkboxType: CheckboxType;
+  filterDebounce = new Subject<string>();
+  toHtmlInputElement = toHtmlInputElement;
+  
+  state: ITreeState;
+  options: ITreeOptions = {
+    useVirtualScroll: true,
+    nodeHeight: 25,
+    displayField: 'name',
+    idField: 'id',
+    allowDrag: false,
+    scrollOnActivate: false,
+    nodeClass: (node: TreeNode) => {
+      const selected = this.selectedOptions.find(option => option.id === node.id);
+      let nodeClass;
+
+      if (!!selected && selected.type === 'excluded') {
+        nodeClass = 'tree-active-exclusion';
+      } else if (!selected && node.isActive) {
+        if (node.parent.getClass().includes('tree-active-exclusion')) {
+          nodeClass = 'tree-active-exclusion tree-active-child';
+        } else {
+          nodeClass = 'tree-active-child';
+        }
+      }
+
+      if (node.isActive && node.isFocused) {
+        return nodeClass + ' tree-active-focused';
+      }
+
+      return nodeClass;
+    },
+    actionMapping: {
+      mouse: {
+        click: (tree: TreeModel, node: TreeNode, $event: any) => {
+          this.nodeToggled(tree, node, $event);
+        }
+      },
+      keys: {
+        [KEYS.SPACE]: (tree: TreeModel, node: TreeNode, $event: any) => {
+          this.nodeToggled(tree, node, $event);
+        },
+        [KEYS.ENTER]: (tree: TreeModel, node: TreeNode, $event: any) => {
+          this.nodeToggled(tree, node, $event);
+        }
+      }
+    }
+  };
+
+  ngOnInit() {
+    this.checkboxType = this.tristate ? CheckboxType.excluded : CheckboxType.basic;
+  }
+
+  constructor(
+  ) {
+    this.filterDebounce.pipe(
+      debounceTime(500)
+    ).subscribe(query => this.filterTree(query));
+  }
+
+  treeInit() {
+    this.treeModel = this.treeComponent.treeModel;
+
+    this.selectedOptions.forEach(selected => {
+      const node = this.treeModel.getNodeById(selected.id);
+
+      if (!node) {
+        return;
+      }
+
+      this.nodeSelected(this.treeModel, node, 'initializing');
+      this.expandParents(this.treeModel, node, null);
+    });
+  }
+
+  onFilterChange(query) {
+    this.filterDebounce.next(query);
+  }
+
+  filterTree(query) {
+    if (query?.length > 0) {
+      this.treeModel.filterNodes(query);
+    } else {
+      this.treeModel.clearFilter();
+      this.treeModel.collapseAll();
+
+      this.selectedOptions.forEach(option => {
+        const node = this.treeModel.getNodeById(option.id);
```

Comment:
in `treeInit` you have a guard for node being falsy, is that needed here? If not, then maybe remove the guard from `treeInit` because now it smells like `this.selectedOptions` can have undefined values and that's a bigger problem that should be solved at it's root.

Reply:
Removed falsy check, should be unnecessary outside truly strange situations.